sigstore/oidc.py

```diff
@@ -0,0 +1,184 @@
+# Copyright 2022 The Sigstore Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+API for retrieving OIDC tokens.
```

Comment:
There's a fair bit of stuff here, but the gist of it is that I've introduced the `detect_credential` and `get_identity_token` functions as well as all the supporting types (exceptions, argument types, etc).

Reply:
Makes sense! It looks like the `get_identity_token` parameters are currently a little "magical", so I'll look into providing reasonable defaults. Ideally, performing the OAuth2 flow _should_ be as simple as:

```python
get_identity_token(Issuer("..."))
```

Reply:
Thinking about it some more, we should really just provide the same producting/staging-style APIs for `Issuer`, i.e.:

```
Issuer.production()
Issuer.staging()
issuer.get_identity_token(...)
```

Reply:
[a8a0a01](https://github.com/sigstore/sigstore-python/pull/383/commits/a8a0a0114c6e9cf1e9ada9bc9187e9629fd7e5b8)